softlayer/resource_softlayer_network_vlan.go

```diff
@@ -0,0 +1,419 @@
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+			"child_resource_count": &schema.Schema{
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+			"subnets": &schema.Schema{
+				Type:     schema.TypeSet,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"subnet": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"subnet_type": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceSoftLayerNetworkVlanCreate(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(*session.Session)
+	router := d.Get("primary_router_hostname").(string)
+	name := d.Get("name").(string)
+
+	vlanType := d.Get("type").(string)
+	if vlanType != "PRIVATE" && vlanType != "PUBLIC" {
+		return fmt.Errorf("Error creating network vlan: vlanType should be either 'PRIVATE' or 'PUBLIC'")
+	}
+	if (vlanType == "PRIVATE" && len(router) > 0 && strings.Contains(router, "fcr")) ||
+		(vlanType == "PUBLIC" && len(router) > 0 && strings.Contains(router, "bcr")) {
+		return fmt.Errorf("Error creating network vlan: mismatch between vlan_type '%s' and primary_router_hostname '%s'", vlanType, router)
+	}
+
+	// Find price items with AdditionalServicesNetworkVlan
+	productOrderContainer, err := buildVlanProductOrderContainer(d, sess, AdditionalServicesNetworkVlanPackageType)
+	if err != nil {
+		// Find price items with AdditionalServices
+		productOrderContainer, err = buildVlanProductOrderContainer(d, sess, AdditionalServicesPackageType)
+		if err != nil {
+			return fmt.Errorf("Error creating network vlan: %s", err)
+		}
+	}
+
+	log.Println("[INFO] Creating network vlan")
+
+	receipt, err := services.GetProductOrderService(sess).
+		PlaceOrder(productOrderContainer, sl.Bool(false))
+	if err != nil {
+		return fmt.Errorf("Error during creation of network vlan: %s", err)
+	}
+
+	vlan, err := findNetworkVlanByOrderId(sess, *receipt.OrderId)
+
+	if len(name) > 0 {
+		_, err = services.GetNetworkVlanService(sess).
+			Id(*vlan.Id).EditObject(&datatypes.Network_Vlan{Name: sl.String(name)})
+		if err != nil {
+			return fmt.Errorf("Error updating Network Vlan: %s", err)
+		}
+	}
+
+	d.SetId(fmt.Sprintf("%d", *vlan.Id))
+	return resourceSoftLayerNetworkVlanRead(d, meta)
+}
+
+func resourceSoftLayerNetworkVlanRead(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(*session.Session)
+	service := services.GetNetworkVlanService(sess)
+
+	vlanId, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return fmt.Errorf("Not a valid vlan ID, must be an integer: %s", err)
+	}
+
+	vlan, err := service.Id(vlanId).Mask(
+		"id," +
+			"name," +
+			"primaryRouter[datacenter[name]]," +
+			"primaryRouter[hostname]," +
+			"vlanNumber," +
+			"billingItem[recurringFee]," +
+			"guestNetworkComponentCount," +
+			"subnets[networkIdentifier,cidr,subnetType]",
```

Comment:
Performance nit. This is better to have as some constant globally in the package.